Add generator for bin based plots for FETs

glatosinski opened the pull request:
Addresses #200 

The scripts generate bin based plots for FETs available in the `sky130_fd_pr` library.

glatosinski commented:
@googlebot I signed it!

glatosinski commented:
My question is - what should be done when there are lots of combinations of W and L parameters? Should I introduce some steps for those values, or limits? We could possibly make some custom legend so it would fit into the plot, but still the plot lines would overlap heavily in some places.

![sim_pfet_01v8_lvt__fT](https://user-images.githubusercontent.com/46489721/99064662-89b3e880-25a6-11eb-9ae3-8805010dc67d.png)

mithro commented:
> My question is - what should be done when there are lots of combinations of W and L parameters? Should I introduce some steps for those values, or limits? We could possibly make some custom legend so it would fit into the plot, but still the plot lines would overlap heavily in some places.

It seems like grouping the plots for different W values (1.0, 3.0, 5.0, 7.0) might work? 

You should ask on the #analog-design channel on slack about what people would find most useful

yrrapt commented:
According to the gm/Id theory you only need to do the plot for one width because you're looking for the current density, drain current/width. So you would sweep the current density for each length to get the plots.

That's the theory anyway, it falls over at the extremes but for the central regions of Id/W it should hold true.

glatosinski commented:
I added grouping plots for the different W values. There is also a possibility to explicitly tell what W values should be considered for plots (in `--only-w` parameter in https://github.com/antmicro/skywater-pdk/blob/200-generate-bin-plots/scripts/python-skywater-pdk/skywater_pdk/fet_simulator/fet_simulator.py or in https://github.com/antmicro/skywater-pdk/blob/200-generate-bin-plots/scripts/python-skywater-pdk/skywater_pdk/fet_simulator/generate-docs-plots.py, where `None` values should be replaced by a list of desired W values).

I also moved legend outside the plot so it does not cover title or plots:

![sim_nfet_01v8__fT_W3_0](https://user-images.githubusercontent.com/46489721/99271843-60ee5600-2827-11eb-95a9-df736480f757.png)

During generation of FET plots it turned out that for some `bins.csv` files the content is invalid, e.g. for https://foss-eda-tools.googlesource.com/skywater-pdk/libs/sky130_fd_pr/+/refs/heads/master/cells/nfet_20v0_nvt_iso/sky130_fd_pr__nfet_20v0_nvt_iso.bins.csv we have:
```
device,bin,W,L
model corner
,0,30.0,1.0
```
instead of 
```
device,bin,W,L
sky130_fd_pr__nfet_20v0_nvt_iso,0,30.0,1.0
```
This occurs for:
https://foss-eda-tools.googlesource.com/skywater-pdk/libs/sky130_fd_pr/+/refs/heads/master/cells/nfet_20v0_nvt_iso/sky130_fd_pr__nfet_20v0_nvt_iso.bins.csv
https://foss-eda-tools.googlesource.com/skywater-pdk/libs/sky130_fd_pr/+/refs/heads/master/cells/nfet_20v0_nvt/sky130_fd_pr__nfet_20v0_nvt.bins.csv
https://foss-eda-tools.googlesource.com/skywater-pdk/libs/sky130_fd_pr/+/refs/heads/master/cells/nfet_20v0_nvt_iso/sky130_fd_pr__nfet_20v0_nvt_iso.bins.csv
https://foss-eda-tools.googlesource.com/skywater-pdk/libs/sky130_fd_pr/+/refs/heads/master/cells/nfet_20v0_zvt/sky130_fd_pr__nfet_20v0_zvt.bins.csv

Later, the bins.csv files are missing for `nfet_20v0` and `nfet_g11v0d16v0` cells - should I use some default values of `bin`, `W` and `L` in such scenarios?

In the end, the PySpice ends successfully but does not provide values for `gm_id`, `id_W`, `ft` and `gm_gds` for the simulation of `pfet_20v0` and `pfet_g5v0d16v0` cells.

mithro commented:
@glatosinski - Please log bugs for those other issues.

mithro commented on scripts/python-skywater-pdk/skywater_pdk/fet_simulator/fet_simulator.py:
```diff
@@ -0,0 +1,209 @@
+        plt.close('all')
+
+
+if __name__ == '__main__':
```
This should be in a `main()` function.

This bit should just be;
```python
if __name__ == '__main__':
    sys.exit(main(sys.argv))
```

glatosinski replied:
fixed

mithro commented on scripts/python-skywater-pdk/skywater_pdk/fet_simulator/generate-docs-plots.py:
```diff
@@ -0,0 +1,72 @@
+
+from fet_simulator import generate_fet_plots
+
+if __name__ == '__main__':
```
This should be in a `main()` function.

This bit should just be;
```python
if __name__ == '__main__':
    sys.exit(main(sys.argv))
```

glatosinski replied:
fixed

mithro commented on scripts/python-skywater-pdk/skywater_pdk/fet_simulator/generate-docs-plots.py:
```diff
@@ -0,0 +1,72 @@
+
+    typicalcorner = args.fd_pr_dir / 'models/corners/tt.spice'
+
+    fets = [
```
Is there any reason this couldn't be auto-generated by just listing the directory with `*fet*`?

glatosinski replied:
When I was writing the script I was relying on the directory names and structure from https://github.com/google/skywater-pdk/tree/master/docs/rules/device-details - the `generate-docs-plots.py` saves SVG files in such a manner that they go to the corresponding FET directories in the Device details. In this directory there were some directories that were named differently in comparison to names in `sky130_fd_pr`, e.g. `nfet_20v0_nvt_iso` from sky130_fd_pr is described in `nfet_20v0_iso` directory, both `nfet_05v0_nvt` and `nfet_03v3_nvt` from sky130_fd_pr are described in `nfet_03v3_vnt-and-nfet_05v0_nvt` directory in Device details.

I can change it so it will scan sky130_fd_pr directory for FET cells and produce directories with FET names and corresponding SVG files in those directories.

mithro replied:
We should fix it so that the device-details directories match the real device names.

glatosinski replied:
I changed the script so it now scans for FET cells in libraries. If the script fails to generate the plot for FET cell it saves the path to the failing bins.csv file to the file given in `--failed-inputs` flag. The SVG files are now saved in such a way that for each:
```
libraries/<library>/<version>/cells/<fet_name>/<library>__<fet_name>.bins.csv
```
file there will be corresponding:
```
output/<library>/<version>/cells/<fet_name>/<library>__<fet_name>__<fet_modelname>_W<width_value>.sim.svg
```
where:
- `library` is the name of the library (FETs are in general located only in `sky130_fd_pr`)
- `fet_name` is FET cell name
- `fet_modelname` is the FET model name (present in Spice files and bins.csv files)
- `width_value` is the width value for which the simulation was performed, with underscore in place of decimal.

The example names are:
```
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_fT_W20_35.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_fT_W40_31.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_fT_W5_4.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_gm_gds_W20_35.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_gm_gds_W40_31.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_gm_gds_W5_4.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_gm_id_W20_35.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_gm_id_W40_31.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_gm_id_W5_4.sim.svg
sky130_fd_pr/latest/cells/esd_nfet_01v8/sky130_fd_pr__esd_nfet_01v8__sky130_fd_pr__esd_nfet_01v8_Id_w_W20_35.sim.svg
```

mithro reviewed (commented):
The plots don't seem to be included in the documentation anywhere?

tclarke commented:
Is this working with the rf_* fet models? When I wrote the original, it did not because these models are actually subcircuits with multiple FETs so you can't get a gm from an internal parameter and likely will have to calculate from the Vgg and Iout. Are gm values composable? (can we combine the individual gm values in the constituent fets?)

mithro commented on scripts/python-skywater-pdk/skywater_pdk/fet_simulator/fet_simulator.py:
```diff
@@ -0,0 +1,293 @@
+#!/usr/bin/env python3
```
Maybe `skywater_pdk/simulation/fet.py` or maybe even `skywater_pdk/simulation/analog/fet.py`?

glatosinski replied:
I used the latter

mithro commented on requirements.txt:
```diff
@@ -4,5 +4,8 @@ flake8
 rst_include
 
+# the PySpice module for rendering FET bins plots
+PySpice==1.4.3
```
Does this version need to be hard coded?

glatosinski replied:
fixed

mithro commented on scripts/python-skywater-pdk/skywater_pdk/fet_simulator/fet_simulator.py:
```diff
@@ -0,0 +1,293 @@
+    plts: plots on which plot lines should be drawn
+    """
+    # plot some interesting things
+    plts[0].plot(gm_id, id_W, label=f'W {fet_W} x L {fet_L}')
```
Maybe `plots` is a better name than `plts`. You are only saving one letter and it took me a second to figure out what this meant.

mithro replied:
It is also very close to the global module `plt`.

glatosinski replied:
fixed

mithro commented on scripts/python-skywater-pdk/skywater_pdk/fet_simulator/fet_simulator.py:
```diff
@@ -0,0 +1,293 @@
+        plt.close(f)
+
+
+def read_bins(fname):
```
Reading bins is something you want to do for multiple reasons (not just simulation). This should probably be part of a generic the `cell` class @ https://github.com/google/skywater-pdk/blob/master/scripts/python-skywater-pdk/skywater_pdk/base.py#L510 ?

glatosinski replied:
Moved the method to Cell class

mithro commented on scripts/python-skywater-pdk/skywater_pdk/fet_simulator/generate-docs-plots.py:
```diff
@@ -0,0 +1,116 @@
+import errno
+
+from fet_simulator import generate_fet_plots
+
```
This module needs a proper doc string (which you can also use in the help).

glatosinski replied:
I added docstrings for:
```
scripts/python-skywater-pdk/skywater_pdk/simulation/analog/fet.py
scripts/python-skywater-pdk/skywater_pdk/simulation/analog/generate-fet-bins-plots.py
```
